### PR TITLE
RD-5759 Migrate app framework - `templateManagement` components to TypeScript

### PR DESCRIPTION
**Pull request by qooban**
## Description
Migrated all remaining files in `app/components/templateManagement` directory to TypeScript.
Minor type-safety improvements included.

## Screenshots / Videos
N/A

## Checklist
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
[Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3877/) (18.11.2022) - failed due to interrupted exception (RD-4319), but all template management related tests passed, so I think we're quite safe;
[Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3878/) (18.11.2022) - PASSED (failing only on known issue - RD-6388)
[Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3884/) (21.11.2022) - in progress

## Documentation
N/A

### CODE REVIEW

File: app/components/templateManagement/templates/TenantList.tsx [hunk 1]
```diff
@@ -25,14 +29,14 @@ export default function TenantList({ custom, onDelete, style, tenants }) {
                                             link
                                             name="remove"
                                             className="right floated"
-                                            onClick={e => e.stopPropagation()}
+                                            onClick={(event: Event) => event.stopPropagation()}
```
Comment (kubama) [inline app/components/templateManagement/templates/TenantList.tsx:32]: Do we need type for `event` to be specified? Can't it be inferred from `onClick` type?

File: app/components/templateManagement/TemplateManagement.tsx [hunk 2]
```diff
@@ -109,34 +115,39 @@ export default function TemplateManagement() {
 
     useEffect(() => {
         dispatch(setTemplateManagementActive(true));
-        return () => dispatch(setTemplateManagementActive(false));
+        return (() => dispatch(setTemplateManagementActive(false))) as () => void;
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/TemplateManagement.tsx:118]: I wasn't digging deep. Faced the following error without casting:

```
TS2345: Argument of type '() => () => SetTemplateManagementActiveAction' is not assignable to parameter of type 'EffectCallback'.   Type '() => SetTemplateManagementActiveAction' is not assignable to type 'void | (() => void | undefined)'.     Type '() => SetTemplateManagementActiveAction' is not assignable to type '() => void | undefined'.       Type 'SetTemplateManagementActiveAction' is not assignable to type 'void'
```

If you have better working suggestion, I'm open to change it.

File: app/components/templateManagement/TemplateManagement.tsx [hunk 3]
```diff
@@ -109,34 +115,39 @@ export default function TemplateManagement() {
...
 
-    function setSelected(collection, id) {
+    function setSelected<Item extends { id: string; selected: boolean }>(collection: Item[], id: string) {
         return _.map(collection, item => ({ ...item, selected: !item.selected && item.id === id }));
     }
 
-    function onSelectTemplate({ id }) {
-        setTemplates(setSelected(templates, id));
+    function onSelectTemplate({ id }: Template) {
+        setTemplates(setSelected<Template>(templates, id));
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/TemplateManagement.tsx]: Sure, removed.

File: app/components/templateManagement/templates/RoleList.tsx [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
-import PropTypes from 'prop-types';
+import { noop } from 'lodash';
 import React from 'react';
 import i18n from 'i18next';
 import { Segment, Icon, Divider, List, Message, PopupConfirm } from '../../basic';
 
-export default function RoleList({ custom, onDelete, roles, style }) {
+interface RoleListProps {
+    custom: boolean;
+    onDelete: (role: string) => any;
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/templates/RoleList.tsx]: Done.

File: app/components/templateManagement/templates/TenantList.tsx [hunk 5]
```diff
@@ -1,22 +1,26 @@
...
 import Const from '../../../utils/consts';
 import { Segment, Icon, Divider, List, Message, PopupConfirm } from '../../basic';
 
-export default function TenantList({ custom, onDelete, style, tenants }) {
+interface TenantListProps {
+    custom: boolean;
+    onDelete: (tenant: string) => any;
+    tenants: string[];
+    style?: React.CSSProperties;
+}
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/templates/TenantList.tsx]: Good idea. Done.

File: app/components/templateManagement/templates/TenantList.tsx (hunk 1 above)
Comment (qooban) [reply app/components/templateManagement/templates/TenantList.tsx:30]: Unfortunately `Icon` doesn't have `onClick` prop defined.

File: app/components/templateManagement/pages/CreatePageModal.tsx [hunk 6]
```diff
@@ -1,26 +1,29 @@
...
+import React from 'react';
 import i18n from 'i18next';
+import type { StrictInputProps } from 'semantic-ui-react';
 import { useBoolean, useErrors, useResettableState } from '../../../utils/hooks';
 import { ApproveButton, Button, CancelButton, Form, Icon, Modal } from '../../basic';
 
-export default function CreatePageModal({ onCreatePage, pageName: initialPageName }) {
-    const [pageName, setPageName] = useState(initialPageName);
+export interface CreatePageModalProps {
+    onCreatePage: (pageName: string) => Promise<any>;
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/pages/CreatePageModal.tsx]: Changed to:
```
    onCreatePage: (pageName: string) => Promise<void | CallHistoryMethodAction>;
```

File: app/components/templateManagement/TemplateManagement.tsx [hunk 7]
```diff
@@ -59,14 +66,15 @@ export default function TemplateManagement() {
...
+                        pages: templateDefs[template.id].pages,
+                        selected: template.id === selectedTemplate?.id
+                    };
                 });
-                if (selectedTemplate) {
-                    (_.find(preparedTemplates, { id: selectedTemplate.id }) || {}).selected = true;
-                }
 
-                const preparedPages = _.map(pageList, page => {
+                const preparedPages = _.map(_.compact(pageList), page => {
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/TemplateManagement.tsx:77]: I'll change import to explicit and will check it deeper next week.

File: app/components/templateManagement/pageGroups/PageGroups.tsx [hunk 8]
```diff
@@ -84,11 +86,11 @@ const PageGroups: FunctionComponent<PageGroupsProps> = ({ pageGroups = [] }) =>
             id: newId,
             name,
             pages: pageIds,
-            icon
+            icon: icon || ''
```
Comment (Vorbert-Kruk) [inline app/components/templateManagement/pageGroups/PageGroups.tsx:89]: WDYT about assigning a default value of `''` to the `icon` argument in the L82

After doing that, we could have

```typescript
        const body = {
            id: newId,
            name,
            pages: pageIds,
            icon
        };
```

Comment (qooban) [reply app/components/templateManagement/pageGroups/PageGroups.tsx:89]: For that I'd need to change `icon` argument type (right now it's `SemanticICONS` which doesn't allow empty string), but I'm not sure that's the way to go here.

File: app/components/templateManagement/TemplateManagement.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (qooban) [reply app/components/templateManagement/TemplateManagement.tsx]: I improved `backend/handler/templates/PagesHandler.ts` and now we shouldn't get nullish items here.

**Review by qooban — COMMENTED** [review]
Ready for re-review.

**Review by kubama — APPROVED** [review]

File: app/components/templateManagement/pageGroups/PageGroups.tsx (hunk 8 above)
Comment (Vorbert-Kruk) [reply app/components/templateManagement/pageGroups/PageGroups.tsx:89]: Oh okay, thanks for the explanation
In that case we should be good to go with leaving it as it is 💪

**Review by Vorbert-Kruk — APPROVED** [review]

**Pull request merged**